src/mono/mono/mini/aot-compiler.c

```diff
@@ -13215,6 +13220,124 @@ add_profile_instances (MonoAotCompile *acfg, ProfileData *data)
 	printf ("Added %d methods from profile.\n", count);
 }
 
+static int
+add_single_mibc_profile_method (MonoAotCompile *acfg, MonoMethod *method)
+{
+	if (!method->is_inflated || mono_method_is_generic_sharable_full (method, FALSE, FALSE, FALSE)) {
+		if (!acfg->aot_opts.profile_only)
+			return 0;
+		if (m_class_get_image (method->klass) != acfg->image)
+			return 0;
+		add_profile_method (acfg, method);
+		return 1;
+	}
+
+	if (acfg->aot_opts.dedup_include) {
+		add_profile_method (acfg, method);
+		return 1;
+	}
+
+	MonoGenericContext *ctx = mono_method_get_context (method);
+	if ((ctx->class_inst && inst_references_image (ctx->class_inst, acfg->image)) ||
+		(ctx->method_inst && inst_references_image (ctx->method_inst, acfg->image))) {
+		add_profile_method (acfg, method);
+		return 1;
+	}
+
+	if (m_class_get_image (method->klass) == acfg->image &&
+		((ctx->class_inst && is_local_inst (ctx->class_inst, acfg->image)) ||
+		(ctx->method_inst && is_local_inst (ctx->method_inst, acfg->image)))) {
+		add_profile_method (acfg, method);
+		return 1;
+	}
+
+	return 0;
+}
+
+static int
+add_mibc_group_method_methods (MonoAotCompile *acfg, MonoMethod *mibcGroupMethod, MonoImage *image, MonoClass *mibcModuleClass, MonoGenericContext *context)
+{
+	ERROR_DECL (error);
+
+	MonoMethodHeader *mibcGroupMethodHeader = mono_method_get_header_internal (mibcGroupMethod, error);
+	mono_error_assert_ok (error);
+
+	int count = 0;
+	uint8_t *cur = (uint8_t*)mibcGroupMethodHeader->code;
+	uint8_t *end = (uint8_t*)mibcGroupMethodHeader->code + mibcGroupMethodHeader->code_size;
+	while (cur < end) {
+		MonoOpcodeEnum il_op;
+		const unsigned char *opcodeIp = (unsigned char*)cur;
+		const unsigned char *opcodeEnd = (unsigned char*)end;
+		cur += mono_opcode_value_and_size (&opcodeIp, opcodeEnd, &il_op);
+		if (il_op != MONO_CEE_LDTOKEN)
+			continue;
+
+		g_assert (opcodeIp + 4 < opcodeEnd);
+		guint32 mibcGroupMethodEntryToken = *(guint32 *)(opcodeIp + 1);
+
+		MonoMethod *methodEntry = mono_get_method_checked (image, mibcGroupMethodEntryToken, mibcModuleClass, context, error);
+		mono_error_assert_ok (error);
```

Comment:
Are these methodref tokens which will resolve to a method in one of the loaded assemblies ?
So we don't need the string based matching used by the current aot profiler ?

Reply:
The methodref tokens are from the .mibc file itself. I lack knowledge of the innerworkings of `mono_get_method_checked` to know exactly how the correct MonoMethod is obtained from just the `.mibc` image, `<Module>` class in that image, and a MonoGenericContext, but it seems to work. Perhaps the .mibc file already embeds information from MethodDetails and BulkType events which provide exactly what methods/types were used?

A breakdown of the .mibc file in the PR description as `Breakdown of HelloWorld Sample .mibc file` you can find in the HelloWorld Assembly mibcGroupMethod, there is one LDTOKEN referencing the `Main` method in the HelloWorld sample, which has a token of `0x0A0002b5`. Looking at the MethodRef section denoted by `(0x0a)`, the `2b5:` index shows the `Main` method.

Reply:
> Are these methodref tokens which will resolve to a method in one of the loaded assemblies ?
> So we don't need the string based matching used by the current aot profiler ?

A mibc file is a "normal" .NET assembly, so it has Method, MemberRef, TypeRef, TypeSpec, AssemblyRef etc tables. 

There are two kinds of things we care about in this file:

1. "mibc group methods" they're in the Method table and have an IL body.  there's a method with a special name which has a body of `ldtoken <groupMethodToken>` - each of those group methods defines profiling data about a collection of assemblies.
2. inside the group methods the bodies have `ldtoken <methodRefToken>` which refers to an actual .NET assembly method like `[S.P.C]System.Object::ToString` or something.  These are like normal memberref tokens - they point to the memberref table in the mibc image and that points to a TypeSpec or a TypeRef and that points to an AssemblyRef etc.

So `mono_get_method_checked (mibcImage, methodRefToken)` will trigger assembly loading of the referenced asembly, initialize the MonoClass that owns the method, etc etc.

Reply:
> So we don't need the string based matching used by the current aot profiler ?

Yea one of the design goals of the mibc format was that it just uses normal .NET lookup for methods that are defined outside the current image.  It's still by name, but using the normal metadata routines